src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -88,11 +88,22 @@ internal AsyncIteratorMethodToStateMachineRewriter(
 
             return (asyncDispatch != null) ? F.Block(asyncDispatch, iteratorDispatch) : iteratorDispatch;
         }
+
+        protected override BoundStatement GenerateHoistedLocalsCleanup(ImmutableArray<StateMachineFieldSymbol> hoistedLocals)
+        {
+            // We need to clean nested hoisted local variables too (not just top-level ones)
```

Comment:
>We need to clean nested hoisted local variables too (not just top-level ones) [](http://example.com/codeflow?start=15&length=77)

This isn't obvious. It looks like there are two call sites of this method and it is not clear why we want to have this behavior for all of them. Instead of completely ignoring the passed argument, consider adjusting the call sites to construct the right set of locals to clear. We can discuss offline in more details.

Reply:
This behavior could be a source of bugs in the future, since the method doesn't do what a reasonable dev would expect it to do.